File: internal/scripts/analyze.py
"""runs a recording of you playing against a song and reports what came out.

same detector as the live worker, so a take judged here and a take judged on
the practice screen agree. one shot: it prints a report event and exits.
"""

import argparse
import json
import sys

import numpy as np
import soundfile as sf

import pitch

# a match is worth more than a wrong note is costly, and a wrong note costs
# less than skipping over both sides, so a fret played wrong lines up with the
# note it was meant to be instead of turning into a miss plus an extra
MATCH = 2.0
MISMATCH = -1.0
GAP = -2.0


def emit(event, message="", data=None):
    line = {"event": event, "message": message}
    if data is not None:
        line["data"] = data
    sys.stdout.write(json.dumps(line) + "\n")
    sys.stdout.flush()


def expected_events(song):
    """collapses the notes into one event per attack, so a chord is one event.

    the detector is monophonic and a strummed chord gives back one pitch out of
    it, whichever rings loudest. counting the chord as six separate demands
    would score every chord in the song as one hit and five misses.
    """
    events = []
    for note in song["notes"]:
        if events and abs(note["time"] - events[-1]["time"]) < 0.02:
            events[-1]["midis"].append(note["midi"])
            events[-1]["frets"].append([note["string"], note["fret"]])
            continue
        events.append(
            {
                "time": note["time"],
                "measure": note["measure"],
                "midis": [note["midi"]],
                "frets": [[note["string"], note["fret"]]],
            }
        )
    return events


def prune(played):
    """drops a reading that was replaced almost at once.

    the first frames of a plucked low string often read an octave down, before
    the fundamental settles, and that reading would be reported as a note the
    player added. nothing on a guitar is played and abandoned in 50 ms
    """
    kept = []
    for index, note in enumerate(played):
        if index + 1 < len(played) and played[index + 1]["t"] - note["t"] < 0.05:
            continue
        kept.append(note)
    return kept


def transcribe(path, progress=True):
    """the notes heard in the file, in order."""
    data, rate = sf.read(path, dtype="float32", always_2d=True)
    mono = data.mean(axis=1)

    tracker = pitch.Tracker(rate)
    played = []
    total = max(1, (len(mono) - pitch.FRAME) // pitch.HOP)
    step = max(1, total // 20)

    for count, start in enumerate(range(0, len(mono) - pitch.FRAME, pitch.HOP)):
        frame = mono[start : start + pitch.FRAME]
        note = tracker.push(frame, start / rate)
        if note is not None:
            played.append(note)
        if progress and count % step == 0:
            emit("progress", data={"done": count, "total": total})

    return prune(played), len(mono) / rate


def matches(midi, expected):
    if midi in expected:
        return True
    # the low strings often read one octave up, because the fundamental of a
    # wound string is quieter than its second harmonic. that is the detector
    # being a detector, not the wrong fret
    return any(abs(midi - value) == 12 for value in expected)


def align(events, played):
    """needleman wunsch over the two note sequences.

    a plain index by index walk breaks on the first extra or skipped note and
    reports every note after it as wrong, which says nothing useful about a
    take where one note was missed in the second measure.
    """
    rows, columns = len(events), len(played)
    scores = np.zeros(columns + 1)
    scores[:] = np.arange(columns + 1) * GAP
    # 0 diagonal, 1 up (an expected note nobody played), 2 left (an extra note)
    pointers = np.zeros((rows + 1, columns + 1), dtype=np.uint8)
    pointers[0, 1:] = 2
    pointers[1:, 0] = 1

    for row in range(1, rows + 1):
        expected = events[row - 1]["midis"]
        hit = np.array(
            [MATCH if matches(note["midi"], expected) else MISMATCH for note in played]
        )

        previous = scores
        scores = np.empty(columns + 1)
        scores[0] = row * GAP

        # the left move reads the cell being written, so this row cannot be
        # vectorized past the diagonal and the up move
        best_diagonal = previous[:-1] + hit
        best_up = previous[1:] + GAP

        for column in range(1, columns + 1):
            diagonal = best_diagonal[column - 1]
            up = best_up[column - 1]
            left = scores[column - 1] + GAP

            if diagonal >= up and diagonal >= left:
                scores[column] = diagonal
                pointers[row, column] = 0
            elif up >= left:
                scores[column] = up
                pointers[row, column] = 1
            else:
                scores[column] = left
                pointers[row, column] = 2

    steps = []
    row, column = rows, columns
    while row > 0 or column > 0:
        move = pointers[row, column]
        if row > 0 and column > 0 and move == 0:
            steps.append(("pair", row - 1, column - 1))
            row -= 1
            column -= 1
        elif row > 0 and move == 1:
            steps.append(("missed", row - 1, None))
            row -= 1
        else:
            steps.append(("extra", None, column - 1))
            column -= 1

    steps.reverse()
    return steps


def report(song, events, played, duration, steps):
    results = []
    hits = 0
    offsets = []

    for kind, index, position in steps:
        if kind == "extra":
            results.append(
                {
                    "kind": "extra",
                    "played": played[position]["name"],
                    "at": played[position]["t"],
                }
            )
            continue

        event = events[index]
        entry = {
            "measure": event["measure"],
            "time": event["time"],
            "frets": event["frets"],
            "expected": [pitch.note_name(midi) for midi in event["midis"]],
        }

        if kind == "missed":
            entry["kind"] = "missed"
        else:
            note = played[position]
            entry["at"] = note["t"]
            entry["played"] = note["name"]
            if matches(note["midi"], event["midis"]):
                entry["kind"] = "hit"
                hits += 1
                offsets.append(note["t"] - event["time"])
            else:
                entry["kind"] = "wrong"

        results.append(entry)

    total = len(events)
    span = events[-1]["time"] - events[0]["time"] if total > 1 else 0.0

    summary = {
        "notes": total,
        "hits": hits,
        "accuracy": round(hits / total, 4) if total else 0.0,
        "extras": sum(1 for step in steps if step[0] == "extra"),
        "missed": sum(1 for step in steps if step[0] == "missed"),
        "duration": round(duration, 2),
    }

    # how fast the take actually was, which is the one number that explains a
    # run where every note is right and it still sounded wrong
    if len(offsets) > 1 and span > 0:
        first, last = offsets[0], offsets[-1]
        summary["tempo"] = round(song["tempo"] * span / (span + last - first), 1)

    by_measure = {}
    for entry in results:
        if entry["kind"] == "extra":
            continue
        bucket = by_measure.setdefault(entry["measure"], {"notes": 0, "hits": 0})
        bucket["notes"] += 1
        bucket["hits"] += entry["kind"] == "hit"

    return {
        "song": song["title"],
        "summary": summary,
        "measures": [
            {"index": index, "notes": value["notes"], "hits": value["hits"]}
            for index, value in sorted(by_measure.items())
        ],
        "notes": results,
    }


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--song", required=True)
    parser.add_argument("--audio", required=True)
    args = parser.parse_args()

    with open(args.song, encoding="utf-8") as handle:
        song = json.load(handle)

    events = expected_events(song)
    if not events:
        emit("analyze_error", "the song has no notes to compare against")
        return 1

    try:
        played, duration = transcribe(args.audio)
    except Exception as error:
        emit("analyze_error", str(error))
        return 1

    if not played:
        emit("analyze_error", "no note was heard in the file")
        return 1

    emit("report", data=report(song, events, played, duration, align(events, played)))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: internal/ui/analyze.go
package ui

import (
	"fmt"
	"strings"

	"github.com/charmbracelet/lipgloss"
)

func (m *Model) viewAnalyze() string {
	if m.asking == askingRecording {
		return m.viewAsk("Listen to a recording",
			"Any wav, flac, ogg or mp3 of you playing the song straight through.\n"+
				"It runs through the same detector the practice screen uses, so the\n"+
				"two agree on what you played.")
	}
	if m.running {
		return m.viewWorking()
	}
	if m.report == nil {
		return m.viewNoReport()
	}

	return m.viewReport()
}

func (m *Model) viewNoReport() string {
	name := "no song picked yet"
	if m.current != nil {
		name = m.current.Title
	}

	lines := []string{
		"",
		m.sectionHead("ANALYZE A RECORDING", name),
		"",
		"  " + styleSubtle.Render("Play the song into a recorder, then let it be marked here."),
		"  " + styleSubtle.Render("Press "+styleAccent.Render("a")+styleSubtle.Render(" and give it the file.")),
		"",
		"  " + styleFaint.Render("It reports the notes you missed, the ones you played instead,"),
		"  " + styleFaint.Render("and how fast the take actually was."),
	}

	return strings.Join(lines, "\n") + blank(m.space()-len(lines))
}

func (m *Model) viewWorking() string {
	lines := []string{
		"",
		m.sectionHead("LISTENING", fmt.Sprintf("%.0f%%", m.progress*100)),
		"",
		"  " + bar(m.progress, m.width-4, styleAccent),
	}

	return strings.Join(lines, "\n") + blank(m.space()-len(lines))
}

func (m *Model) viewReport() string {
	summary := m.report.Summary
	style := styleOk
	switch {
	case summary.Accuracy < 0.6:
		style = styleBad
	case summary.Accuracy < 0.85:
		style = styleWarn
	}

	lines := []string{
		"",
		pad("  "+styleHeading.Render(m.report.Song),
			style.Render(fmt.Sprintf("%.0f%%", summary.Accuracy*100))+
				styleFaint.Render(fmt.Sprintf("  of %d notes", summary.Notes)), m.width),
		"",
		"  " + m.counts(summary),
		"",
		"  " + styleAccent.Render("BY MEASURE"),
		"",
	}

	lines = append(lines, m.measureRows()...)
	lines = append(lines, "", "  "+styleAccent.Render("WHAT TO FIX"), "")
	lines = append(lines, m.problemRows(m.space()-len(lines)-2)...)

	return strings.Join(lines, "\n")
}

func (m *Model) counts(summary reportSummary) string {
	parts := []string{
		styleOk.Render(fmt.Sprint(summary.Hits)) + styleFaint.Render(" right"),
		styleBad.Render(fmt.Sprint(summary.Notes-summary.Hits-summary.Missed)) + styleFaint.Render(" wrong"),
		styleWarn.Render(fmt.Sprint(summary.Missed)) + styleFaint.Render(" missed"),
		styleSubtle.Render(fmt.Sprint(summary.Extras)) + styleFaint.Render(" extra"),
	}

	line := strings.Join(parts, styleFaint.Render("    "))
	if summary.Tempo > 0 {
		line += styleFaint.Render("    played at ") + styleSubtle.Render(fmt.Sprintf("♩%.0f", summary.Tempo))
	}

	return line
}

// measureRows is the small bar chart that says where the trouble is. Reading
// four hundred verdicts one by one says nothing; one row per measure says the
// second half of the solo is the part to go back to.
func (m *Model) measureRows() []string {
	const cells = 12

	var rows []string
	line := "  "
	for index, measure := range m.report.Measures {
		share := 0.0
		if measure.Notes > 0 {
			share = float64(measure.Hits) / float64(measure.Notes)
		}

		style := styleOk
		switch {
		case share < 0.6:
			style = styleBad
		case share < 0.85:
			style = styleWarn
		}

		block := styleFaint.Render(fmt.Sprintf("%3d ", measure.Index)) +
			style.Render(strings.Repeat("▆", int(share*cells))) +
			styleFaint.Render(strings.Repeat("▁", cells-int(share*cells)))

		line += block + "   "
		if (index+1)%4 == 0 {
			rows = append(rows, strings.TrimRight(line, " "))
			line = "  "
			if len(rows) >= 6 {
				break
			}
		}
	}

	if strings.TrimSpace(line) != "" {
		rows = append(rows, strings.TrimRight(line, " "))
	}

	return rows
}

// problemRows lists only what went wrong. A list that also holds every note
// that was right is a list nobody scrolls to the end of.
func (m *Model) problemRows(room int) []string {
	if room < 1 {
		return nil
	}

	var problems []reportNote
	for _, note := range m.report.Notes {
		if note.Kind != "hit" {
			problems = append(problems, note)
		}
	}

	if len(problems) == 0 {
		return []string{"  " + styleOk.Render("nothing. every note landed.")}
	}

	start := m.reportRow
	if start > len(problems)-1 {
		start = len(problems) - 1
	}
	end := start + room
	if end > len(problems) {
		end = len(problems)
	}

	rows := make([]string, 0, end-start)
	for _, note := range problems[start:end] {
		rows = append(rows, m.problemRow(note))
	}

	if end < len(problems) {
		rows[len(rows)-1] = styleFaint.Render(fmt.Sprintf("  %d more", len(problems)-end+1))
	}

	return rows
}

// position writes a note the way a tab writes it inline, string letter then
// fret. Printing the string as a number reads as a fraction and nobody can
// tell which half is which.
func (m *Model) position(str, fret int) string {
	letter := fmt.Sprint(str)
	if m.current != nil {
		letter = m.current.Label(str)
	}
	return fmt.Sprintf("%s|%d", letter, fret)
}

func (m *Model) problemRow(note reportNote) string {
	where := styleFaint.Render(fmt.Sprintf("  m%-4d", note.Measure))

	var frets []string
	for _, fret := range note.Frets {
		frets = append(frets, m.position(fret[0], fret[1]))
	}

	expected := styleSubtle.Render(strings.Join(note.Expected, " ")) +
		styleFaint.Render("  "+strings.Join(frets, " "))

	verdict := styleWarn.Render("not played")
	if note.Kind == "wrong" {
		verdict = styleBad.Render("played " + note.Played)
	}
	if note.Kind == "extra" {
		expected = styleFaint.Render("nothing written here")
		verdict = styleSubtle.Render("played " + note.Played)
	}

	left := where + expected
	return pad(truncate(left, m.width-lipgloss.Width(verdict)-4), verdict+"  ", m.width)
}
